### PR TITLE
AFR (and most of AFC) Tokens

### PR DESCRIPTION
**Pull request by mainman879**
Tokens added for (so far): Prosperous Innkeeper, Lolth, Spider Queen, Drizzt Do'Urden, and Ellywick Tumblestrum. 

Could not find the pictures for these tokens on Scryfall but will add picurls to them when they are added. I have not added any dungeon cards or their relations as I do not know whether they should use the token relation system or not.

### CODE REVIEW

**Comment by tooomm** [comment]
In order to correctly flag this as <kbd>WIP</kbd> and to prevent an accidental merge before it's ready, I converted this into a `Draft PR`.
Just hit the `Ready for review` button when you are done! @mainman879 👍

**Comment by mainman879** [comment]
@tooomm what do you think about merging this before the final few images get leaked for upcoming tokens (which could take weeks). There's a ton of functionality in these changes that people want now (dungeons mainly). Adding the last few token images can be done later very easily (I will be opening a new PR for the AFC set after this one is merged and can include those last images in it too.)

**Comment by tooomm** [comment]
There are another two weeks before the set officially releases. I think they probably add the other pics by then.

However, with things being available on MTGA and people wanting to play the set already...
It makes sense to have something useful in between I guess? Some people might complain about missing stuff though.

As this is likely to be the new normal for upcoming releases we can give it a try maybe? @Psithief

**Comment by mainman879** [comment]
The rest of the tokens for AFR are now on scryfall. I'll be adding them later today and setting this as ready for review.

File: tokens.xml [hunk 1]
```diff
@@ -3931,8 +4016,8 @@ When this creature enters the battlefield, it deals 1 damage to any target.</tex
             <reverse-related>Battle for Bretagard</reverse-related>
             <reverse-related>Beskir Shieldmate</reverse-related>
             <reverse-related>Maja, Bretagard Protector</reverse-related>
-            <reverse-related exclude="exclude" count="x">Maja, Bretagard Protector</reverse-related>
-            <reverse-related count="x=1">Usher of the Fallen</reverse-related>
+            <reverse-related count="x" exclude="exclude">Maja, Bretagard Protector</reverse-related>
+            <reverse-related>Usher of the Fallen</reverse-related>
```
Comment (tooomm) [inline tokens.xml:4020]: @Psithief added the `count="x=1"` in https://github.com/Cockatrice/Magic-Token/commit/f0b7fee4eb1ccd9dbf2f23a0d9b53acd7ed173d1#diff-ef2c5b2e0216513ba7314de2c6637a3b48b09a5ea3ee44d13c951195fb293697R3716

Maybe the idea was to account for something like double Boast triggers?

But looking at other Boast cards that create tokens they also have no count value. 👍 
https://scryfall.com/search?q=o%3Aboast+o%3Atoken

File: tokens.xml [hunk 2]
```diff
@@ -7098,6 +7247,21 @@ This creature can't be enchanted.</text>
...
+            <text>Deathtouch</text>
+            <prop>
+                <colors>B</colors>
+                <type>Token Legendary Creature — God</type>
+                <maintype>Creature</maintype>
+                <cmc>0</cmc>
+                <pt>4/4</pt>
+            </prop>
+            <set picURL="https://c1.scryfall.com/file/scryfall-cards/large/front/6/5/65f8e40f-fb5e-4ab8-add3-a8b87e7bcdd9.jpg?1626139916">AFR</set>
+            <!-- this token is created by another token only -->
```
Comment (tooomm) [inline tokens.xml:7261]: Why not reverse-link to the `Tomb of Annihilation` dungeon token here?

Instead you linked it the other way around :) https://github.com/Cockatrice/Magic-Token/pull/108/files#diff-ef2c5b2e0216513ba7314de2c6637a3b48b09a5ea3ee44d13c951195fb293697R8783

**Review by tooomm — COMMENTED** [review]
(Conflicts with the master branch are resolved.)

Gave this a brief look, I did not double check if there is a token missing or if all card relations are correct.

I assume the empty set tags like `<set>AFC</set>` are placeholders which do miss card art links for now?

Please update the date at the beginning of the file (and in the [dedicated version file](https://github.com/mainman879/Magic-Token/blob/patch-1/version.txt)) as well!
(Forgot this myself as well when doing the quick beast fix the other day... 🙈 )

File: tokens.xml (hunk 2 above)
Comment (mainman879) [reply tokens.xml:7261]: I did it as related because there is precedence for it in other tokens such as the Ajani emblem that is related to cat tokens. There are a few others as well. It doesn't make a difference either way I suppose but we should keep it consistent one way or the other.

File: tokens.xml (hunk 1 above)
Comment (mainman879) [reply tokens.xml:4020]: I think the count for this relation should be removed as Boast is once per turn, and none of the other token relations assume a token doubler or similar.

**Comment by mainman879** [comment]
The empty AFC set tags are for AFC specific tokens that haven't been revealed yet. These can be added later, and mostly serve as a reminder for myself to fill them in when possible.

File: tokens.xml (hunk 1 above)
Comment (tooomm) [reply tokens.xml:4020]: >I think the count for this relation should be removed as Boast is once per turn, and none of the other token relations assume a token doubler or similar.

Agreed.

File: tokens.xml [hunk 3]
```diff
@@ -3156,6 +3225,7 @@ Whenever Glyph Keeper becomes the target of a spell or ability for the first tim
             <reverse-related count="3">Siege-Gang Commander</reverse-related>
             <reverse-related count="2">Sling-Gang Lieutenant</reverse-related>
             <reverse-related>Survey the Wreckage</reverse-related>
+            <reverse-related count="2">You See a Pair of Goblins</reverse-related>
```
Comment (Psithief) [inline tokens.xml:3228]: Battle Cry Goblin creates 1/1 red goblins
Den of the Bugbear creates 1/1 red goblins

File: tokens.xml [hunk 4]
```diff
@@ -8438,6 +8654,173 @@ Sacrifice this creature: This creature deals 1 damage to target creature or play
             <token>1</token>
             <tablerow>2</tablerow>
         </card>
+        <card>
+            <name>Dungeon of the Mad Mage</name>
+            <text>Yawning Portal — You gain 1 life. (→ Dungeon Level)
+Dungeon Level — Scry 1. (→ Goblin Bazaar or Twisted Caverns)
+Goblin Bazaar — Create a Treasure token. (→ Lost Level)
```
Comment (Psithief) [inline tokens.xml:8661]: The treasure token relation is missing.

File: tokens.xml [hunk 5]
```diff
@@ -4086,6 +4187,21 @@ When this creature enters the battlefield, it deals 1 damage to any target.</tex
             <token>1</token>
             <tablerow>2</tablerow>
         </card>
+        <card>
+            <name>Illusion    </name>
```
Comment (Psithief) [inline tokens.xml:4191]: You must add another space at the end of this token's name, because the previous token already ends in four spaces.

File: tokens.xml [hunk 6]
```diff
@@ -8262,7 +8476,9 @@ Sacrifice this creature: This creature deals 1 damage to target creature or play
             <reverse-related>Varina, Lich Queen</reverse-related>
             <reverse-related>Vile Rebirth</reverse-related>
             <reverse-related>Wakedancer</reverse-related>
+            <reverse-related count="x">Wand of Orcus</reverse-related>
```
[the remark this answers is not in the document]
Comment (Psithief) [reply tokens.xml:8479]: The Book of Vile Darkness also creates 2/2 Zombies.

File: tokens.xml [hunk 7]
```diff
@@ -8458,8 +8841,8 @@ Sacrifice this creature: This creature deals 1 damage to target creature or play
                 <maintype>Emblem</maintype>
             </prop>
             <set picURL="https://img.scryfall.com/cards/large/front/1/c/1c97e5b2-a024-4aa7-a9b8-45f441aad138.jpg">M19</set>
-            <reverse-related>Ajani, Adversary of Tyrants</reverse-related>
             <related count="3">Cat   </related>
+            <reverse-related>Ajani, Adversary of Tyrants</reverse-related>
```
Comment (Psithief) [inline tokens.xml:8845]: Reordering elements should be a separate pull request.

File: tokens.xml [hunk 8]
```diff
@@ -8438,6 +8654,173 @@ Sacrifice this creature: This creature deals 1 damage to target creature or play
...
+Runestone Caverns — Exile the top two cards of your library. You may play them. (→ Deep Mines)
+Muiral’s Graveyard — Create two 1/1 black Skeleton creature tokens. (→ Deep Mines)
+Deep Mines — Scry 3. (→ Mad Wizard’s Lair)
+Mad Wizard’s Lair — Draw three cards and reveal them. You may cast one of them without paying its mana cost.</text>
+            <prop>
+                <type>Dungeon</type>
+                <maintype>Dungeon</maintype>
+            </prop>
+            <set picURL="https://c1.scryfall.com/file/scryfall-cards/large/front/6/f/6f509dbe-6ec7-4438-ab36-e20be46c9922.jpg?1625106684">AFR</set>
+            <related count="2" exclude="exclude">Skeleton </reverse-related>
```
Comment (Psithief) [inline tokens.xml:8673]: You have mismatched start and end tags here. This breaks the parsing for the rest of the file.

**Review by Psithief — CHANGES_REQUESTED** [review]

**Comment by Psithief** [comment]
Add this URL for `Dragon Spirit `: https://media.wizards.com/2021/afr/en_7iaWmuoF0c.png

File: tokens.xml [hunk 9]
```diff
@@ -2713,6 +2765,21 @@ When Earthshaker Khenra enters the battlefield, target creature with power less
             <token>1</token>
             <tablerow>2</tablerow>
         </card>
+        <card>
+            <name>Faerie Dragon</name>
```
[the remark this answers is not in the document]
Comment (mainman879) [reply tokens.xml:2769]: Fixed.

File: tokens.xml [hunk 10]
```diff
@@ -8132,6 +8342,8 @@ Sacrifice this creature: This creature deals 1 damage to target creature or play
                 <cmc>0</cmc>
                 <pt>2/2</pt>
             </prop>
+            <set>AFC</set>
```
[the remark this answers is not in the document]
Comment (mainman879) [reply tokens.xml:8345]: Removed all the AFC set tags except for the new illusion and the dragon spirit, and filled in their picurls.

File: tokens.xml (hunk 4 above)
Comment (mainman879) [reply tokens.xml:8661]: It is already right below the Skeleton relation.

File: tokens.xml (hunk 3 above)
Comment (mainman879) [reply tokens.xml:3228]: Fixed.

File: tokens.xml [hunk 11]
```diff
@@ -7291,54 +7458,78 @@ This creature can't be enchanted.</text>
...
+            <reverse-related>Spiked Pit Trap</reverse-related>
             <reverse-related>Storm the Vault</reverse-related>
             <reverse-related>Storm-Kiln Artist</reverse-related>
             <reverse-related>Strike It Rich</reverse-related>
             <reverse-related>Sudden Breakthrough</reverse-related>
             <reverse-related count="x">Surly Badgersaur</reverse-related>
+            <reverse-related count="x">Swarming Goblins</reverse-related>
             <reverse-related count="2">Tavern Scoundrel</reverse-related>
             <reverse-related>Tempting Contract</reverse-related>
+            <reverse-related>Thieve's Tools</reverse-related>
```
[the remark this answers is not in the document]
Comment (mainman879) [reply tokens.xml:7520]: Fixed.

File: tokens.xml [hunk 12]
```diff
@@ -7291,54 +7458,78 @@ This creature can't be enchanted.</text>
...
             <reverse-related>Smashing Success</reverse-related>
             <reverse-related>Smothering Tithe</reverse-related>
             <reverse-related count="x">Spell Swindle</reverse-related>
+            <reverse-related>Spiked Pit Trap</reverse-related>
             <reverse-related>Storm the Vault</reverse-related>
             <reverse-related>Storm-Kiln Artist</reverse-related>
             <reverse-related>Strike It Rich</reverse-related>
             <reverse-related>Sudden Breakthrough</reverse-related>
             <reverse-related count="x">Surly Badgersaur</reverse-related>
+            <reverse-related count="x">Swarming Goblins</reverse-related>
```
[the remark this answers is not in the document]
Comment (mainman879) [reply tokens.xml:7517]: Fixed.

File: tokens.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mainman879) [reply tokens.xml:8479]: Fixed.

**Comment by mainman879** [comment]
I do not want to delete the sorting and putting things into a consistent order just to immediately reupload it under 2 other pull requests. If someone wants to do this and edit this PR and reupload it themselves I welcome them to do so. I'll keep those conversations as unresolved.

**Comment by mainman879** [comment]
Thanks for fixing the tab issues @tooomm . I usually check for this issue manually but must've forgot in my morning grogginess. I'll have to get an actual code editor instead of notepad++ because it seems to love to use tab indents instead of spaces.

**Comment by Psithief** [comment]
I'll me reviewing and/or merging this in a few hours time.

**Comment by Psithief** [comment]
I extracted and committed 1b781c26f09bce468c9c6769c3dca596206d19df in your name.

**Comment by Psithief** [comment]
@mainman879 I propose to squash this pull request as per the commits shown here: https://github.com/Cockatrice/Magic-Token/commits/pullrequest108

File: tokens.xml [hunk 13]
```diff
@@ -3130,6 +3199,7 @@ Whenever Glyph Keeper becomes the target of a spell or ability for the first tim
             <reverse-related count="x">Goblin Goliath</reverse-related>
             <reverse-related>Goblin Instigator</reverse-related>
             <reverse-related count="2">Goblin Marshal</reverse-related>
+            <reverse-related attach="attach">Goblin Morningstar</reverse-related>
```
Comment (Psithief) [inline tokens.xml:3202]: Goblin Morningstar does not attach 45% of the time, so there should be another relation to create a goblin without the attach attribute.
Even though this is less likely to happen, it should probably still be the default (i.e. `exclude` the `attach` relation) otherwise players will need to manually detach the equipment from the token if they use the default action.

File: tokens.xml [hunk 14]
```diff
@@ -6153,6 +6272,21 @@ Equip {1}</text>
...
+            <name>Skeleton </name>
+            <prop>
+                <colors>B</colors>
+                <type>Token Creature — Skeleton</type>
+                <maintype>Creature</maintype>
+                <cmc>0</cmc>
+                <pt>1/1</pt>
+            </prop>
+            <set picURL="https://c1.scryfall.com/file/scryfall-cards/large/front/f/a/fa6fdb57-82f3-4695-b1fa-1f301ea4ef83.jpg?1626139846">AFR</set>
+            <reverse-related>Skeletal Swarming</reverse-related>
```
Comment (Psithief) [inline tokens.xml:6285]: Death-Priest of Myrkul creates a skeleton.

**Review by Psithief — CHANGES_REQUESTED** [review]
I found two issues. I was sure I mentioned the Death-Priest but apparently not.

**Comment by Psithief** [comment]
> It makes sense to have something useful in between I guess? Some people might complain about missing stuff though.
> 
> As this is likely to be the new normal for upcoming releases we can give it a try maybe? @Psithief

What about a spoiler branch that's not quite held to the same standard as the master branch? @tooomm

File: tokens.xml (hunk 13 above)
Comment (mainman879) [reply tokens.xml:3202]: Added

File: tokens.xml (hunk 14 above)
Comment (mainman879) [reply tokens.xml:6287]: Added

**Comment by mainman879** [comment]
> 
> 
> @mainman879 I propose to squash this pull request as per the commits shown here: https://github.com/Cockatrice/Magic-Token/commits/pullrequest108

I don't know how to squash or what squashing is on github but whatever you think is good go for it.

**Review by Psithief — APPROVED** [review]

**Comment by ebbit1q** [comment]
I thought you intended to squash them? now we have these commits applied to master directly

**Comment by Psithief** [comment]
@ebbit1q The commits with the same author got squashed in sequence. I would have liked to have squashed it further but would have lost authorship information. If it there was just one author I'd have squashed it to one commit.
image

**Comment by ebbit1q** [comment]
whatever makes you happy, I don't care much about the small fixes I do
